tests/PHPStan/Analyser/data/preg_match_shapes.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace PregMatchShapes;
+
+use function PHPStan\Testing\assertType;
+
+function doMatch(string $s): void {
+	if (preg_match('/Price: (£|€)\d+/', $s, $matches)) {
+		assertType('array{0: string, 1?: string}', $matches);
+	}
+	assertType('array<string>', $matches);
+
+	if (preg_match('/Price: (£|€)(\d+)/i', $s, $matches)) {
+		assertType('array{0: string, 1?: string, 2?: string}', $matches);
+	}
+	assertType('array<string>', $matches);
+
+	if (preg_match('/(a)(b)*(c)(d)*/', $s, $matches)) {
+		assertType('array{0: string, 1?: string, 2?: string, 3?: string, 4?: string}', $matches);
+	}
+	assertType('array<string>', $matches);
+}
+
+function doNonCapturingGroup(string $s): void {
+	if (preg_match('/Price: (?:£|€)(\d+)/', $s, $matches)) {
+		assertType('array{0: string, 1?: string}', $matches);
+	}
+	assertType('array<string>', $matches);
+}
+
+function doNamedSubpattern(string $s): void {
+	if (preg_match('/\w-(?P<num>\d+)-(\w)/', $s, $matches)) {
+		assertType('array{0: string, num?: string, 1?: string, 2?: string}', $matches);
+	}
+	assertType('array<string>', $matches);
+}
+
+function doOnlyNmedSubpattern(string $s): void {
+	// n modifier captures only named groups
+	if (preg_match('/(\w)-(?P<num>\d+)-(\w)/n', $s, $matches)) {
+		assertType('array<string>', $matches); // could be "array{0: string, num?: string, 1?: string}"
```

Comment:
this case does not work with the trick from https://3v4l.org/sOXbn - therefore added it to show the current result as I had to early exit in the `RegexShapeMatcher` for it

Reply:
same trick with last, but named, capture group works - https://3v4l.org/3SdDM